### PR TITLE
Interpret timeout zero value as no limit

### PR DESCRIPTION
**Pull request by jack-berg**
Implementation of spec PR: https://github.com/open-telemetry/opentelemetry-specification/pull/4331

Supersedes #6850.

### CODE REVIEW

File: exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpGrpcSender.java [hunk 1]
```diff
@@ -81,11 +81,15 @@ public OkHttpGrpcSender(
       @Nullable RetryPolicy retryPolicy,
       @Nullable SSLContext sslContext,
       @Nullable X509TrustManager trustManager) {
+    int callTimeoutMillis =
+        (int) Math.min(Duration.ofNanos(timeoutNanos).toMillis(), Integer.MAX_VALUE);
+    int connectTimeoutMillis =
+        (int) Math.min(Duration.ofNanos(connectTimeoutNanos).toMillis(), Integer.MAX_VALUE);
```
Comment (jack-berg) [inline exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpGrpcSender.java:87]: Internally, OkHttp converts durations in int type and time unit millis. `Duration` allows us to specify a value that overflows OkHttp's internal logic, so we need to add some bounds.

**Review by breedx-splk — APPROVED** [review]
Looks good. I had a small question about negative values and maybe see an issue with MAX_VALUE for Integer in okhttp...but otherwise seems like a good change.

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporterBuilder.java [hunk 2]
```diff
@@ -87,7 +87,7 @@ public GrpcExporterBuilder<T> setChannel(ManagedChannel channel) {
   }
 
   public GrpcExporterBuilder<T> setTimeout(long timeout, TimeUnit unit) {
-    timeoutNanos = unit.toNanos(timeout);
+    timeoutNanos = timeout == 0 ? Long.MAX_VALUE : unit.toNanos(timeout);
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporterBuilder.java:90]: > Do we want to think about negative values while we're at it...or leave for a follow-up effort?

Negative values are rejected upstream of this, e.g.: https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/otlp/all/src/main/java/io/opentelemetry/exporter/otlp/http/trace/OtlpHttpSpanExporterBuilder.java#L81

We discussed whether to centralize this type of parameter validation in a central place or duplicate it many times in upper layers, and decided on duplicating in upper layers. The idea being, that its much more discoverable for users to discover the requirements of the parameter if the validation is directly in the builder, vs. having to step through an obscure chain of abstractions to reach GrpcExporterBuilder.

We could of course have the checks at multiple layers, but that comes with its own problems of duplication and more edge cases to test if you take a strict view of coverage.

File: exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpGrpcSender.java [hunk 3]
```diff
@@ -81,11 +81,15 @@ public OkHttpGrpcSender(
...
     OkHttpClient.Builder clientBuilder =
         new OkHttpClient.Builder()
             .dispatcher(OkHttpUtil.newDispatcher())
-            .callTimeout(Duration.ofNanos(timeoutNanos))
-            .connectTimeout(Duration.ofNanos(connectTimeoutNanos));
+            .callTimeout(Duration.ofMillis(callTimeoutMillis))
+            .connectTimeout(Duration.ofMillis(connectTimeoutMillis));
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpGrpcSender.java:92]: > Unfortunately, it looks like the Integer.MAX_VALUE that this clips to above is also a problem. Maybe just clip to MAX_VALUE-1?

I don't follow - could you elaborate?

I was curious to make sure things work at the edges, so I add some additional unit tests testing at the limits using various versions of `Long.MAX_VALUE` in [this commit](). Notes:

- `setTimeout(Long.MAX_VALUE, TimeUnit.NANOSECONDS)` is valid
- `setTimeout(Long.MAX_VALUE, TimeUnit.SECONDS)` is valid - `TimeUnit#toNanos` is used to convert under the hood, and its smart enough to not overlow `Long`. 
- `setTimeout(Duration.ofNanos(Long.MAX_VALUE))` is valid
- `setTimeout(Duration.ofSeconds(Long.MAX_VALUE))` is _not_ valid and generates `ArithmeticException`. It seems that even with the java std library, the overflow behavior is inconsistent.

[the remark this answers is not in the document]
Comment (breedx-splk) [reply exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpGrpcSender.java:92]: Hmmm, odd that it's passing.

I was looking at `OkHttpClientBuilder.setTimeout()` which calls `Util.checkDuration()` https://github.com/square/okhttp/blob/9c653c7b0a3c63774a57e4e1c0e07e4f67ca1333/okhttp/src/commonJvmAndroid/kotlin/okhttp3/internal/-UtilJvm.kt#L132 which looks like it should throw if the value is equal to MAX_VALUE. 

BUT I'm seeing it now....the `Math.min` is using `Integer.MAX_VALUE` but the okhttp require checks against `Long.MAX_VALUE`....so we're fine. It just wasn't obvious to me when reading that. Sorry for the distraction.

[the remark this answers is not in the document]
Comment (jack-berg) [reply exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpGrpcSender.java:92]: Here's the OkHttp require check source code:

>   check(!duration.isNegative()) { "$name < 0" }
  val millis = duration.inWholeMilliseconds
  require(millis <= Integer.MAX_VALUE) { "$name too large" }
  require(millis != 0L || !duration.isPositive()) { "$name too small" }
  return millis.toInt()

It does check against `Integer.MAX_VALUE`, but its a `<=` and not a `<`, which is why clipping to `Integer.MAX_VALUE` works.
